### ISSUE TITLE
H2: Issues remaining with SHUTR/SHUTW and RST_STREAM

### ISSUE THREAD

**Issue by wtarreau** [haproxy/haproxy#192]
## Output of `haproxy -vv` and `uname -a`

```
$ uname -a
Linux wtap.local 4.19.48 #2 SMP Sat Jun 8 10:06:46 CEST 2019 x86_64 GNU/Linux
$ ../haproxy/haproxy  -v
HA-Proxy version 2.1-dev1-d6e0c0-118 2019/07/25 - https://haproxy.org/
$ ../haproxy-2.0/haproxy  -v
HA-Proxy version 2.0.3 2019/07/23 - https://haproxy.org/
$ ../haproxy-1.9/haproxy  -v
HA-Proxy version 1.9.9 2019/07/23 - https://haproxy.org/
```

## What's the configuration?

```
defaults
        mode http
        #log stdout daemon
        #option httplog
        timeout client 50s
        timeout server  5s
        timeout connect 5s

listen h2h1
        bind :4446 proto h2
        http-request redirect location /redirect if { path_beg /r }
        http-request deny
        server httpterm 127.0.0.1:8000
```

## Steps to reproduce the behavior

Sending this command shows a lot of REFUSED_STREAMs in 2.0 and above and in 1.9 it freezes the transfer because the RST_STREAM is not emitted:

```
$ while :; do curl -v --http2-prior-knowledge http://127.0.0.1:4446/ -o /dev/null -d @CHANGELOG; done
```

## Actual behavior

1.9: transfer freezing (missing RST_STREAM)
2.0/2.1: RST_STREAM code REFUSED_STREAM (encourages to retry instead of CANCEL which doesn't want it anymore).

## Expected behavior

Unclear. Possibly that we should drain the whole data until the client aborts seeing it gets its respones. But on the other hand the upper layer has really cancelled so sending CANCEL seems suited. At least we must not freeze nor send a code suggesting the request is retryable.

## Do you have any idea what may have caused this?

In 2.0, patch f983d00a1 ("BUG/MINOR: mux-h2: make the do_shut{r,w} functions more robust against retries") solved the freeze that still happens in 1.9. It can relatively easily be adapted to 1.9 to do the same but it remains wrong. The problem is that the HLOC state is excluded from the first block of h2_do_shutw(), making it handled only in the second block, leading to REFUSED_STREAM to be emitted. It's worth noting that this situation matches none of the conditions described by the comments there. Also it doesn't make sense to proceed again with HLOC in shutw(): it's supposed to have been done already. Only for shutr() it makes sense.

## Do you have an idea how to solve the issue?

At the very least, aborting the transfer and not sending REFUSED_STREAM seem mandatory.  This part needs to be restored at the top of h2_do_shutw():
```
	if (h2s->st == H2_SS_HLOC)
		return 0;
```
And at the same time the explicit tests for this state should be removed. In h2_do_shutr() we must specially deal with this case with this:
```
@@ -3165,6 +3172,12 @@ static int h2_do_shutr(struct h2s *h2s)
                 */
                h2s_error(h2s, H2_ERR_REFUSED_STREAM);
        }
+       else {
+               /* a final response was already provided, we don't want this
+                * stream anymore.
+                */
+               h2s_error(h2s, H2_ERR_CANCEL);
+       }
 
        if (!(h2s->flags & H2_SF_RST_SENT) &&
            h2s_send_rst_stream(h2c, h2s) <= 0)
```
In 1.9 this also requires porting part of the patch mentioned above, which consists in not immediately returning on HLOC. Note that 1.9 doesn't have H2_SF_WANT_SHUT{R,W} that some later patches use, so some care is needed.

It seems again that part of the problem here is that "error" is a competing state for our streams and that instead we should let the state always match the protocol state machine and have a separate error flag which could be combined with these.

**Comment by wtarreau** [comment]
It goes a bit further than this. We're even seeing RST_STREAM produced by the following sequence:
  - event on the connection / GOAWAY / etc
  - wake_one_stream() called with CS_FL_ERROR or anything
  - wake_one_stream() unconditionally changes the stream's state to H2_SS_ERROR but does not set an errcode
  - the next frame that arrives (e.g. DATA) doesn't match the expected states, and causes h2s_error(REFUSED_STREAM) to be called, and since the error was not set prior to this, it gets assigned.

We can avoid the transition to H2_SS_ERROR in wake_one_stream() but what about the other ones? Can we decide that a closed connection finishes transitions just like an ES bit would do? Doesn't seem so. But it's likely that the whole shutr/shutw reporting logic still heavily relies on this due to inheritage from 1.8's original design.

**Comment by capflam** [comment]
1.9 is EOL

**Comment by wtarreau** [comment]
It looks like we addressed this in 2.1-dev2 with commits 231f61617, 082c45769 and cfba9d6ea, which were backported to 2.0. Some improvements are still needed in this area but the issue described here is definitely fixed.
